upup/pkg/fi/cloudup/awstasks/iamrole.go

```diff
@@ -143,6 +143,87 @@ func (s *IAMRole) CheckChanges(a, e, changes *IAMRole) error {
 }
 
 func (_ *IAMRole) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *IAMRole) error {
+	if e.RolePolicyDocument == nil {
+		klog.V(2).Infof("Deleting IAM role %q", a.Name)
+
+		var attachedPolicies []*iam.AttachedPolicy
+		var policyNames []string
+
+		// List Inline policies
+		{
+			request := &iam.ListRolePoliciesInput{
+				RoleName: a.Name,
+			}
+			err := t.Cloud.IAM().ListRolePoliciesPages(request, func(page *iam.ListRolePoliciesOutput, lastPage bool) bool {
+				for _, policy := range page.PolicyNames {
+					policyNames = append(policyNames, aws.StringValue(policy))
+				}
+				return true
+			})
+			if err != nil {
+				if awsup.AWSErrorCode(err) == iam.ErrCodeNoSuchEntityException {
+					klog.V(2).Infof("Got NoSuchEntity describing IAM RolePolicy; will treat as already-deleted")
+					return nil
+				}
+
+				return fmt.Errorf("error listing IAM role policies: %v", err)
+			}
+		}
+
+		// List Attached Policies
+		{
+			request := &iam.ListAttachedRolePoliciesInput{
+				RoleName: a.Name,
+			}
+			err := t.Cloud.IAM().ListAttachedRolePoliciesPages(request, func(page *iam.ListAttachedRolePoliciesOutput, lastPage bool) bool {
+				attachedPolicies = append(attachedPolicies, page.AttachedPolicies...)
+				return true
+			})
+			if err != nil {
+				if awsup.AWSErrorCode(err) == iam.ErrCodeNoSuchEntityException {
+					klog.V(2).Infof("Got NoSuchEntity describing IAM RolePolicy; will treat as already-detached")
+					return nil
+				}
+
+				return fmt.Errorf("error listing IAM role policies for %v", err)
+			}
+		}
+
+		// Delete inline policies
+		for _, policyName := range policyNames {
+			klog.V(2).Infof("Deleting IAM role policy %q", policyName)
+			request := &iam.DeleteRolePolicyInput{
+				RoleName:   a.Name,
+				PolicyName: aws.String(policyName),
+			}
+			_, err := t.Cloud.IAM().DeleteRolePolicy(request)
+			if err != nil {
+				return fmt.Errorf("error deleting IAM role policy %q: %v", policyName, err)
+			}
+		}
+
+		// Detach Managed Policies
+		for _, policy := range attachedPolicies {
+			klog.V(2).Infof("Detaching IAM role policy %q", policy)
+			request := &iam.DetachRolePolicyInput{
+				RoleName:  a.Name,
+				PolicyArn: policy.PolicyArn,
+			}
+			_, err := t.Cloud.IAM().DetachRolePolicy(request)
+			if err != nil {
+				return fmt.Errorf("error detaching IAM role policy %q: %v", *policy.PolicyArn, err)
+			}
+		}
```

Comment:
I wonder if it works to `UpdateAssumeRolePolicy` to a PolicyDocument that grants nothing, but I consider that future cleanup work.